experiments/replicaset.go

```diff
@@ -0,0 +1,200 @@
+
+	newRS := appsv1.ReplicaSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            experimentutil.ReplicasetNameFromExperiment(experiment, template),
+			Namespace:       experiment.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(experiment, controllerKind)},
+			Labels:          newRSTemplate.Labels,
+		},
+		Spec: appsv1.ReplicaSetSpec{
+			Replicas:        new(int32),
+			MinReadySeconds: template.MinReadySeconds,
+			Selector:        newRSSelector,
+			Template:        newRSTemplate,
+		},
+	}
+
+	newReplicasCount := experimentutil.CalculateTemplateReplicasCount(experiment, template)
+	*(newRS.Spec.Replicas) = newReplicasCount
+
+	// Create the new ReplicaSet. If it already exists, then we need to check for possible
+	// hash collisions. If there is any other error, we need to report it in the status of
+	// the Rollout.
+	alreadyExists := false
+	createdRS, err := ec.kubeclientset.AppsV1().ReplicaSets(experiment.Namespace).Create(&newRS)
+	switch {
+	// We may end up hitting this due to a slow cache or a fast resync of the Rollout.
+	case errors.IsAlreadyExists(err):
+		alreadyExists = true
+
+		// Fetch a copy of the ReplicaSet.
+		rs, rsErr := ec.replicaSetLister.ReplicaSets(newRS.Namespace).Get(newRS.Name)
+		if rsErr != nil {
+			return nil, rsErr
+		}
+
+		// If the Rollout owns the ReplicaSet and the ReplicaSet's PodTemplateSpec is semantically
+		// deep equal to the PodTemplateSpec of the Rollout, it's the Rollout's new ReplicaSet.
+		// Otherwise, this is a hash collision and we need to increment the collisionCount field in
+		// the status of the Rollout and requeue to try the creation in the next sync.
+		controllerRef := metav1.GetControllerOf(rs)
+		if controllerRef != nil && controllerRef.UID == experiment.UID && replicasetutil.PodTemplateEqualIgnoreHash(&rs.Spec.Template, &template.Template) {
+			createdRS = rs
+			err = nil
+			break
+		}
+
+		templateStatusPtr, statusIndex := experimentutil.GetTemplateStatus(experiment, template)
+		templateStatus := v1alpha1.TemplateStatus{
+			Name: template.Name,
+		}
+		if templateStatusPtr != nil {
+			templateStatus = *templateStatusPtr
+		}
+
+		// Matching ReplicaSet is not equal - increment the collisionCount in the RolloutStatus
+		// and requeue the Rollout.
+		if templateStatus.CollisionCount == nil {
+			templateStatus.CollisionCount = new(int32)
+		}
+		preCollisionCount := *templateStatus.CollisionCount
+		*templateStatus.CollisionCount++
+
+		if statusIndex != nil {
+			experiment.Status.TemplateStatuses[*statusIndex] = templateStatus
+		} else {
+			experiment.Status.TemplateStatuses = append(experiment.Status.TemplateStatuses, templateStatus)
+		}
+
+		// Update the collisionCount for the Rollout and let it requeue by returning the original
+		// error.
+		_, roErr := ec.arogProjClientset.ArgoprojV1alpha1().Experiments(experiment.Namespace).Update(experiment)
+		if roErr == nil {
```

Comment:
I think we should at least log update error. 

Also is it possible to use patch in this case to reduce the possibility of concurrent update error?